### DISCUSSION TITLE
How can I measure the performance impact of L4 vs L7 mode of LoxiLB load balancer, when it is deployed in-cluster with 5G Core ?

### DISCUSSION THREAD

**Discussion by souravpaul8** [loxilb-io/loxilb#791] [Q&A] [answered]
I want to measure, CPU usage and memory usage when LoxiLB is deployed inside Kubernetes in L4 and L7 mode. 
Please give me some direction on how to go about it ?

In L4, I understand that I would need to measure the metrics for loxilb pod, which does the actual load balancing. But in L7 mode as given in theis article:
https://www.loxilb.io/post/ngap-load-balancing-with-loxilb

Only kube-loxilb is present. No loxilb pod. Does L7 only work in external mode where the docker container running outside cluster does the L7 load balancing ?

**Comment by backguynn** [comment]
Hello. First, for loxilb L7 loadbalancing, loxilb in-cluster mode also supports it.

**Comment by TrekkieCoder** [comment]
sctp L7 loadbalancing has not yet been tested in-cluster because we need to integrate with kubernetes ingress api. This is a work in progress. Tcp(http/https) L7 LB however is available currently in-cluster.

Regarding L4 load-balancing in-cluster mode, it is difficult to measure exact cpu utilization by getting pod stats since it uses eBPF running in kernel space. You can try to.give [this](https://github.com/Netflix/bpftop) a try.

**Reply by souravpaul8** [reply]
There are numerous ebpf programs LoxiLB. How do I get the total cpu usage of the loxiLb L4 LB ?
<img width="1194" alt="Screenshot 2024-09-11 at 8 20 08 PM" src="https://github.com/user-attachments/assets/1e902df1-bfb7-474e-b038-e6a7e7f17ffd">

**Reply by TrekkieCoder** [reply]
Hi @souravpaul8  _tc_packet_func__  is the entry point which does most of the heavy lifting. But it can branch out to other tail-called eBPF programs in certain cases. So, for exact measurements , you might want to sum it all up.

**Reply by souravpaul8** [reply]
Okay got it

**Comment by UltraInstinct14** [comment]
Usually L4 LB is deployed infront of L7 LB. But if there is no such requirement, you can deploy loxilb in-cluster and use either L4 or L7 as per your need. In L4 mode, you need to use eBPF specific tool to measure CPU usage etc like previous comment. For L7 mode, currently SCTP LB does not use eBPF so you can use normal Kubernetes specific tools to check pod CPU usage.

**Reply by souravpaul8** [reply]
This [blog](https://www.loxilb.io/post/ngap-load-balancing-with-loxilb) describes how to deploy SCTP NGAP load balancer in a docker container outside the cluster. How to deploy it inside my Kubernetes cluster ?

**Reply by TrekkieCoder** [reply]
It is running as a docker container in the blog because it load-balances across Open5GS running in two separate clusters. There is another blog explaining how to run [in-cluster](https://www.loxilb.io/post/k8s-nuances-of-in-cluster-external-service-lb-with-loxilb) . Probably you are already following that for L4 mode. In the same loxilb deployment yaml, you could change loxilb image which supports L7 SCTP. PS - However it will only load-balancer in the same cluster in this case.

**Reply by souravpaul8** [reply]
Yes, I followed the blog you linked for the L4 Mode. 
So just changing the image in loxilb yaml:
ghcr.io/loxilb-io/loxilb:latest -> http://ghcr.io/loxilb-io/loxilb:scp
and changing the service annotation:
L4:
annotations:
    loxilb.io/lbmode: "fullnat"
    loxilb.io/epselect : "rr"
L7:
annotations:
    http://loxilb.io/lbmode : "fullproxy"
    http://loxilb.io/epselect: “n2”
    
@TrekkieCoder Is this correct ?

**Reply by TrekkieCoder** [reply]
@souravpaul8 Yes, it should work. If it does not, will take further look.

**Reply by souravpaul8** [reply]
@TrekkieCoder How can I be verify, that it is indeed working in L7 mode ?

**Reply by TrekkieCoder** [reply]
You can check the following in the cluster:

```
kubectl exec -it -n kube-system <loxilb-lb-xxx> -- loxicmd get lb - o wide
kubectl exec -it -n kube-system <loxilb-lb-xxx> -- loxicmd get ct
```
You can get loxilb pod ids from:

```
kubectl get pods -A | grep loxilb
```
 If the above commmds show some entries related to the ongoing connection, that would mean it is working.

**Reply by souravpaul8** [reply]
@TrekkieCoder  On what NGAP/NAS property, are you doing the L7 Load Balancing ?

**Reply by TrekkieCoder** [reply]
Currently, loxilb uses RAN UE NGAP ID  for L7 load-balancing.  It is currently built to work with NGAP ASN.1 APER encoding.

**Reply by souravpaul8** [reply]
```
kubectl exec -it -n kube-system <loxilb-lb-xxx> -- loxicmd get lb - o wide
kubectl exec -it -n kube-system <loxilb-lb-xxx> -- loxicmd get ct
```

What will be the difference in the entries of these pods, in L7 mode compared to the L4 mode ?

**Reply by TrekkieCoder** [reply]
## For L4 mode, you will get something like the following - 

```
root@025629b9ac79:/# loxicmd get lb -o wide
|   EXT IP   | SEC IPS | HOST | PORT | PROTO | NAME | MARK | SEL |  MODE   |  ENDPOINT  | EPORT | WEIGHT | STATE |     COUNTERS      |
|------------|---------|------|------|-------|------|------|-----|---------|------------|-------|--------|-------|-------------------|
| 20.20.20.1 |         |      | 2020 | sctp  |      |    0 | rr  | fullnat | 31.31.31.1 |  8080 |      1 | -     | 1760214:246376192 |
|            |         |      |      |       |      |      |     |         | 32.32.32.1 |  8080 |      1 | -     | 1360214:128234561 |
root@025629b9ac79:/# loxicmd get ct        
| SERVICE NAME |    DESTIP    |    SRCIP     | DPORT | SPORT | PROTO | STATE |                 ACT                 | PACKETS |   BYTES   |
|--------------|--------------|--------------|-------|-------|-------|-------|-------------------------------------|---------|-----------|
|              | 20.20.20.1   | 10.10.10.1   |  2020 | 42413 | sctp  | est   | fdnat-20.20.20.1,31.31.31.1:8080:w0 |  119811 | 336045178 |
|              | 20.20.20.1   | 31.31.31.1   | 42413 |  8080 | sctp  | est   | fsnat-20.20.20.1,10.10.10.1:2020:w0 |  159751 |   9904780 |
```

## For L7 mode with NGAP parsing enabled, you will get something like the following - 

```
root@025629b9ac79:/# loxicmd get lb -o wide
|    EXT IP    | SEC IPS | HOST | PORT | PROTO | NAME | MARK | SEL |   MODE    |  ENDPOINT  | EPORT | WEIGHT | STATE | COUNTERS |
|--------------|---------|------|------|-------|------|------|-----|-----------|------------|-------|--------|-------|----------|
| 10.10.10.254 |         |      | 2020 | sctp  |      |    0 | n2  | fullproxy | 31.31.31.1 |  8080 |      1 | -     | 12:12    |
|              |         |      |      |       |      |      |     |           | 32.32.32.1 |  8080 |      1 | -     | 12:12    |
root@025629b9ac79:/# loxicmd get ct
| SERVICE NAME |    DESTIP    |   SRCIP    | DPORT | SPORT | PROTO | STATE |                     ACT                     | PACKETS | BYTES |
|--------------|--------------|------------|-------|-------|-------|-------|---------------------------------------------|---------|-------|
|              | 10.10.10.254 | 10.10.10.1 |  2020 | 10011 | sctp  | est   | fp|31.31.31.254:48612->31.31.31.1:8080|sctp |      24 |    24 |
|              |              |            |       |       |       |       | fp|32.32.32.254:41336->32.32.32.1:8080|sctp |         |       |
```

Further netstat in loxilb, will show the L7 connections -

```
root@025629b9ac79:/# netstat -an | grep "2020\|8080"
sctp                10.10.10.254:2020                               LISTEN     
sctp       0      0 10.10.10.254:2020       10.10.10.1:10011        ESTABLISHED
sctp       0      0 31.31.31.254:48612      31.31.31.1:8080         ESTABLISHED
sctp       0      0 31.31.31.254:41336      32.32.32.1:8080         ESTABLISHED
```

**Reply by souravpaul8** [reply]
My Service's Exteranal IP remains in pending state. 
And this command --- loxicmd get lb -o wide
returns empty.
Is there any problem with the config ?
Does Load Balancing work on accessing the service from within the cluster using cluster IP ?
My cluster is deployed on my own bare metal server, connected to a private network.
I have Flannel and Multus installed.
 - --externalCIDR=<Private_Network_Node_IP>/16
 - --setBGP=64512
 -  - --setRoles=0.0.0.0
<img width="831" alt="Screenshot 2024-09-23 at 8 21 05 PM" src="https://github.com/user-attachments/assets/a41cc576-e416-4ca3-b54b-b1f3e9ef73f6">

**Reply by TrekkieCoder** [reply]
You can get some clue by doing - 
```
kubectl logs -n kube-system --selector app=kube-loxilb-app
```
If you share the above logs, it will help understand the issue.

**Reply by TrekkieCoder** [reply]
I think this is due to the fact, kube-loxilb needs some additional permissions now to support [in-cluster multus](https://github.com/loxilb-io/loxilbdocs/blob/main/docs/loxilb-incluster-multus.md). 

```  
diff --git a/manifest/in-cluster/kube-loxilb.yaml b/manifest/in-cluster/kube-loxilb.yaml
index fb2d8ef..7629825 100644
--- a/manifest/in-cluster/kube-loxilb.yaml
+++ b/manifest/in-cluster/kube-loxilb.yaml
@@ -33,6 +33,7 @@ rules:
     resources:
       - endpoints
       - services
+      - namespaces
       - services/status
```

If you add the above line to your kube-loxilb.yaml and reapply , the errors should go away.

**Reply by souravpaul8** [reply]
@TrekkieCoder  Yes adding namespaces, solved the issue, now got the Ip

**Reply by souravpaul8** [reply]
@TrekkieCoder I am testing the L4 mode. Now External IP and Load Balancing is working fine for TCP connection, for example nginx.
But for SCTP, it is accessible from inside the cluster using cluster IP but doesn't respond, when accessed from outside the cluster using External IP.  I have used the sctp.yaml from https://github.com/loxilb-io/loxilb/blob/main/cicd/k3s-flannel-incluster/sctp.yml
Same is happening with the connection between gNB and AMF for SCTP connection.  I also deployed the LB in L7 mode, in that SCTP connection is getting aborted and resulting in Connection Reset by Peer.
![Screenshot 2024-09-23 at 10 24 50 PM](https://github.com/user-attachments/assets/b05f5f62-77cb-47f9-87b5-bf3ed73ed3c4)
<img width="968" alt="Screenshot 2024-09-23 at 10 48 24 PM" src="https://github.com/user-attachments/assets/07b446a6-c75f-45a9-b4ed-0843136b9a88">

One more question, does Load Balancing work on accessing the service from within the cluster using cluster IP ?

**Reply by TrekkieCoder** [reply, marked as answer]
For sctp L4 LB, you need to do install another loxilb module in the k8s nodes for in-cluster SCTP LB to work :

```
curl -sfL https://github.com/loxilb-io/loxilb-ebpf/raw/main/kprobe/install.sh | sh -
```

For L7, I am not sure what might be the exact problem.

**Reply by souravpaul8** [reply]
@TrekkieCoder Thanks, This solves the issue with the L4 Load Balancer

**Comment by prudentcircle** [comment]
@souravpaul8 If you are curious how much CPU,Mem,Network a Pod uses, my suggestion would be to use Prometheus along with Kube-state-metrics. 


There are very active projects like Kube-Prometheus-Stack that lets you quickly install Prometheus and kube-state-metrics in one helm command. 

Once you do that, you can find out how much a Pod is using CPU, MEM, and Network.
